### ISSUE TITLE
UUID Normalization

### ISSUE THREAD

**Issue by rgbkrk** [ipython/ipykernel#473]
Today I was looking into sticking messages into a database when I noticed that our `msg_id` and `session_id` for `jupyter console` and `ipython` don't match the canonical format (e.g. `3a9993f9-ba0c-4a75-bb58-62af0ac3d2d4`):

```
$ jupyter console --kernel python3
Jupyter console 6.0.0

Python 3.7.6 (default, Dec 30 2019, 19:38:26)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.6.1 -- An enhanced Interactive Python. Type '?' for help.



In [1]: ip = get_ipython()

In [2]: ip.parent_header
Out[2]:
{'header': {'msg_id': '46806cf5-b41b930ae1b3f209d20bcc8c',
  'msg_type': 'execute_request',
  'username': 'kylek',
  'session': '97790200-f0614d420db48cbb8394735e',
  'date': datetime.datetime(2020, 1, 6, 19, 50, 45, 482522, tzinfo=tzutc()),
  'version': '5.3'},
 'msg_id': '46806cf5-b41b930ae1b3f209d20bcc8c',
 'msg_type': 'execute_request',
 'parent_header': {},
 'metadata': {},
 'content': {'code': 'ip.parent_header',
  'silent': False,
  'store_history': True,
  'user_expressions': {},
  'allow_stdin': True,
  'stop_on_error': True},
 'buffers': []}

In [3]: %connect_info
{
  "shell_port": 53739,
  "iopub_port": 53740,
  "stdin_port": 53741,
  "control_port": 53742,
  "hb_port": 53743,
  "ip": "127.0.0.1",
  "key": "c49fffbb-dfb77272de2bdb4133453432",
  "transport": "tcp",
  "signature_scheme": "hmac-sha256",
  "kernel_name": ""
}

Paste the above JSON into a file, and connect with:
    $> jupyter <app> --existing <file>
or, if you are local, you can connect with just:
    $> jupyter <app> --existing kernel-11139.json
or even just:
    $> jupyter <app> --existing
if this is the most recent Jupyter kernel you have started.
```

Is this intended? I realize that I can just convert these all to 128 bit integers for the sake of Postgres. Perhaps the following is a jupyter_client / protocol question though -- should we document that these are string based IDs and are sensitive to `-` placement?

**Comment by minrk** [comment]
From [the spec](https://jupyter-client.readthedocs.io/en/stable/messaging.html#general-message-format):

> 'msg_id' : str, # typically UUID, must be unique per message
> 'session' : str, # typically UUID, should be unique per session

According to the spec, these are only required to be unique strings, their contents and structure are not meaningful. The exact string format is not dictated, but uuids are a common and simple way of accomplishing this. Every uuid-generating form is a valid msg id, but multiple equivalent string representations of the same uuid are not equivalent ids in the protocol (`'a-b-c-d' != 'abcd' != 'ABCD'`).  A persistent single uuid plus a local counter, for instance, would be another valid way to make unique msg ids without generating a uuid for each msg. I thought I remembered something about an upper character limit being requested by @SylvainCorlay for static memory allocations? I can't find that, though.

**Comment by SylvainCorlay** [comment]
>  I thought I remembered something about an upper character limit being requested by @SylvainCorlay for static memory allocations? I can't find that, though.

Yes, in xeus, we assume a max length for the uuid strings so that we can allocate uuids on the stack (and not the heap) which helps performance. I think there is a documentation PR in jupyter-client about this but I am AFK at the moment.

**Comment by SylvainCorlay** [comment]
See #253.
